File: Python3/41. 数据流中的中位数.py
class MedianFinder:

    def __init__(self):
        """
        initialize your data structure here.
        """
        self.maxHeap = []
        self.minHeap = []

    def addNum(self, num: int) -> None:
        # 为了使数据平均分配到两堆：若两堆的数据数目相等，就让 minHeap 的元素个数多1
        # 具体做法分两步：
        # 1、将新元素加入到 maxHeap；
        # 2、将 maxHeap 的堆顶元素加入到 minHeap
        # Python中默认为最小堆，使用最大堆的方法参考40
        if len(self.maxHeap) == len(self.minHeap):
            heappush(self.minHeap, -heappushpop(self.maxHeap, -num))
        else:
            heappush(self.maxHeap, -heappushpop(self.minHeap, num))

    def findMedian(self) -> float:
        if len(self.maxHeap) == len(self.minHeap):
            return (self.minHeap[0]-self.maxHeap[0])/2.0
        else:
            return self.minHeap[0]

File: C++/41. 数据流中的中位数.cpp
class MedianFinder {
   public:
    /** initialize your data structure here. */
    MedianFinder() {
    }

    void addNum(int num) {
        // 为了使数据平均分配到两堆：
        // 一、若两堆的数据数目相等，则让 minHeap 的元素个数 + 1
        if (maxHeap.size() == minHeap.size()) {
            // 1、将新元素加入到 maxHeap；
            maxHeap.push(num);
            // 2、将 maxHeap 的堆顶元素加入到 minHeap
            minHeap.push(maxHeap.top());
            // 3、弹出 maxHeap 的堆顶元素
            maxHeap.pop();
        } else {
            // 二、若不相等，则让 maxHeap 的元素个数 + 1
            minHeap.push(num);
            maxHeap.push(minHeap.top());
            minHeap.pop();
        }
    }

    double findMedian() {
        if (maxHeap.size() == minHeap.size()) {
            return (maxHeap.top() + minHeap.top()) / 2.0;
        }
        return minHeap.top();
    }

   private:
    // C++中优先队列默认为“最大”堆：
    // 元素按照 compare 给出的函数排序，“最大”（即最后面）的元素放到堆顶
    priority_queue<int, vector<int>, less<int>> maxHeap;     // maxHeap 放较小的一半数字
    priority_queue<int, vector<int>, greater<int>> minHeap;  // minHeap 放较大的一半数字
};

/**
 * Your MedianFinder object will be instantiated and called as such:
 * MedianFinder* obj = new MedianFinder();
 * obj->addNum(num);
 * double param_2 = obj->findMedian();
 */